From 231d014789e6633c7c14c80b32f330948da2dd93 Mon Sep 17 00:00:00 2001
From: Bibo Mao <maobibo@loongson.cn>
Date: Fri, 6 Jun 2025 14:30:27 +0800
Subject: [PATCH 204/285] hw/intc/loongarch_pch: Add kernel irqchip realize
 function

Function kvm_pic_realize() is added if kvm_irqchip_in_kernel() return true.
It is to notify KVM kernel to create and initialize PCH PCI device in
kernel mode.

Reviewed-by: Song Gao <gaosong@loongson.cn>
Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Message-ID: <20250606063033.2557365-8-maobibo@loongson.cn>
Signed-off-by: Song Gao <gaosong@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>
Signed-off-by: Binbin Zhou <zhoubinbin@loongson.cn>
---
 hw/intc/Kconfig                     |   3 -
 hw/intc/loongarch_pch_pic.c         |  14 ++-
 hw/intc/loongarch_pch_pic_kvm.c     | 184 ----------------------------
 hw/intc/loongarch_pic_kvm.c         |  38 ++++++
 hw/intc/meson.build                 |   3 +-
 hw/loongarch/Kconfig                |   1 -
 hw/loongarch/virt.c                 |  23 +---
 include/hw/intc/loongarch_pch_pic.h |  37 +-----
 8 files changed, 57 insertions(+), 246 deletions(-)
 delete mode 100644 hw/intc/loongarch_pch_pic_kvm.c
 create mode 100644 hw/intc/loongarch_pic_kvm.c

diff --git a/hw/intc/Kconfig b/hw/intc/Kconfig
index e074c06834f6..efd779ae0265 100644
--- a/hw/intc/Kconfig
+++ b/hw/intc/Kconfig
@@ -105,9 +105,6 @@ config LOONGARCH_PCH_PIC
     bool
     select UNIMP
 
-config LOONGARCH_PCH_PIC_KVM
-    bool
-
 config LOONGARCH_PCH_MSI
     select MSI_NONBROKEN
     bool
diff --git a/hw/intc/loongarch_pch_pic.c b/hw/intc/loongarch_pch_pic.c
index cb63f797580a..1c3a000008de 100644
--- a/hw/intc/loongarch_pch_pic.c
+++ b/hw/intc/loongarch_pch_pic.c
@@ -10,6 +10,7 @@
 #include "qemu/log.h"
 #include "hw/irq.h"
 #include "hw/intc/loongarch_pch_pic.h"
+#include "system/kvm.h"
 #include "trace.h"
 #include "qapi/error.h"
 #include "sysemu/kvm.h"
@@ -284,10 +285,15 @@ static void loongarch_pic_realize(DeviceState *dev, Error **errp)
 
     qdev_init_gpio_out(dev, s->parent_irq, s->irq_num);
     qdev_init_gpio_in(dev, pch_pic_irq_handler, s->irq_num);
-    memory_region_init_io(&s->iomem, OBJECT(dev),
-                          &loongarch_pch_pic_ops,
-                          s, TYPE_LOONGARCH_PIC, VIRT_PCH_REG_SIZE);
-    sysbus_init_mmio(sbd, &s->iomem);
+
+    if (kvm_irqchip_in_kernel()) {
+        kvm_pic_realize(dev, errp);
+    } else {
+        memory_region_init_io(&s->iomem, OBJECT(dev),
+                              &loongarch_pch_pic_ops,
+                              s, TYPE_LOONGARCH_PIC, VIRT_PCH_REG_SIZE);
+        sysbus_init_mmio(sbd, &s->iomem);
+    }
 }
 
 static void loongarch_pic_class_init(ObjectClass *klass, void *data)
diff --git a/hw/intc/loongarch_pch_pic_kvm.c b/hw/intc/loongarch_pch_pic_kvm.c
deleted file mode 100644
index 5e5bba1e4dab..000000000000
--- a/hw/intc/loongarch_pch_pic_kvm.c
+++ /dev/null
@@ -1,184 +0,0 @@
-/* SPDX-License-Identifier: GPL-2.0-or-later */
-/*
- * LoongArch kvm pch pic interrupt support
- *
- * Copyright (C) 2024 Loongson Technology Corporation Limited
- */
-
-#include "qemu/osdep.h"
-#include "hw/qdev-properties.h"
-#include "qemu/typedefs.h"
-#include "hw/intc/loongarch_pch_pic.h"
-#include "hw/sysbus.h"
-#include "linux/kvm.h"
-#include "migration/vmstate.h"
-#include "qapi/error.h"
-#include "sysemu/kvm.h"
-#include "hw/loongarch/virt.h"
-#include "hw/pci-host/ls7a.h"
-#include "qemu/error-report.h"
-
-static void kvm_pch_pic_access_regs(int fd, uint64_t addr,
-                                       void *val, int is_write)
-{
-        kvm_device_access(fd, KVM_DEV_LOONGARCH_PCH_PIC_GRP_REGS,
-                          addr, val, is_write, &error_abort);
-}
-
-static int kvm_loongarch_pch_pic_pre_save(void *opaque)
-{
-    KVMLoongArchPCHPIC *s = (KVMLoongArchPCHPIC *)opaque;
-    KVMLoongArchPCHPICClass *class = KVM_LOONGARCH_PCH_PIC_GET_CLASS(s);
-    int fd = class->dev_fd;
-
-    kvm_pch_pic_access_regs(fd, PCH_PIC_MASK_START,
-                            (void *)&s->int_mask, false);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_HTMSI_EN_START,
-                            (void *)&s->htmsi_en, false);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_EDGE_START,
-                            (void *)&s->intedge, false);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_AUTO_CTRL0_START,
-                            (void *)&s->auto_crtl0, false);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_AUTO_CTRL1_START,
-                            (void *)&s->auto_crtl1, false);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_ROUTE_ENTRY_START,
-                            (void *)s->route_entry, false);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_HTMSI_VEC_START,
-                            (void *)s->htmsi_vector, false);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_INT_IRR_START,
-                            (void *)&s->intirr, false);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_INT_ISR_START,
-                            (void *)&s->intisr, false);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_POLARITY_START,
-                            (void *)&s->int_polarity, false);
-
-    return 0;
-}
-
-static int kvm_loongarch_pch_pic_post_load(void *opaque, int version_id)
-{
-    KVMLoongArchPCHPIC *s = (KVMLoongArchPCHPIC *)opaque;
-    KVMLoongArchPCHPICClass *class = KVM_LOONGARCH_PCH_PIC_GET_CLASS(s);
-    int fd = class->dev_fd;
-
-    kvm_pch_pic_access_regs(fd, PCH_PIC_MASK_START,
-                            (void *)&s->int_mask, true);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_HTMSI_EN_START,
-                            (void *)&s->htmsi_en, true);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_EDGE_START,
-                            (void *)&s->intedge, true);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_AUTO_CTRL0_START,
-                            (void *)&s->auto_crtl0, true);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_AUTO_CTRL1_START,
-                            (void *)&s->auto_crtl1, true);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_ROUTE_ENTRY_START,
-                            (void *)s->route_entry, true);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_HTMSI_VEC_START,
-                            (void *)s->htmsi_vector, true);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_INT_IRR_START,
-                            (void *)&s->intirr, true);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_INT_ISR_START,
-                            (void *)&s->intisr, true);
-    kvm_pch_pic_access_regs(fd, PCH_PIC_POLARITY_START,
-                            (void *)&s->int_polarity, true);
-
-    return 0;
-}
-
-static void kvm_pch_pic_handler(void *opaque, int irq, int level)
-{
-    if (kvm_enabled()) {
-        kvm_set_irq(kvm_state, irq, !!level);
-    }
-}
-
-static void kvm_loongarch_pch_pic_realize(DeviceState *dev, Error **errp)
-{
-    KVMLoongArchPCHPICClass *pch_pic_class =
-            KVM_LOONGARCH_PCH_PIC_GET_CLASS(dev);
-    struct kvm_create_device cd = {0};
-    uint64_t pch_pic_base = VIRT_PCH_REG_BASE;
-    Error *err = NULL;
-    int ret;
-
-    pch_pic_class->parent_realize(dev, &err);
-    if (err) {
-        error_propagate(errp, err);
-        return;
-    }
-
-    if (!pch_pic_class->is_created) {
-        cd.type = KVM_DEV_TYPE_LOONGARCH_PCHPIC;
-        ret = kvm_vm_ioctl(kvm_state, KVM_CREATE_DEVICE, &cd);
-        if (ret < 0) {
-            error_setg_errno(errp, errno,
-                             "Creating the KVM pch pic device failed");
-            return;
-        }
-        pch_pic_class->is_created = true;
-        pch_pic_class->dev_fd = cd.fd;
-        fprintf(stdout, "Create LoongArch pch pic irqchip in KVM done!\n");
-
-        ret = kvm_device_access(cd.fd, KVM_DEV_LOONGARCH_PCH_PIC_GRP_CTRL,
-                                KVM_DEV_LOONGARCH_PCH_PIC_CTRL_INIT,
-                                &pch_pic_base, true, NULL);
-        if (ret < 0) {
-            error_report(
-                "KVM EXTIOI: failed to set the base address of EXTIOI");
-            exit(1);
-        }
-
-        qdev_init_gpio_in(dev, kvm_pch_pic_handler, VIRT_PCH_PIC_IRQ_NUM);
-    }
-}
-
-static const VMStateDescription vmstate_kvm_loongarch_pch_pic = {
-    .name = TYPE_LOONGARCH_PCH_PIC,
-    .version_id = 1,
-    .minimum_version_id = 1,
-    .pre_save = kvm_loongarch_pch_pic_pre_save,
-    .post_load = kvm_loongarch_pch_pic_post_load,
-    .fields = (const VMStateField[]) {
-        VMSTATE_UINT64(int_mask, KVMLoongArchPCHPIC),
-        VMSTATE_UINT64(htmsi_en, KVMLoongArchPCHPIC),
-        VMSTATE_UINT64(intedge, KVMLoongArchPCHPIC),
-        VMSTATE_UINT64(intclr, KVMLoongArchPCHPIC),
-        VMSTATE_UINT64(auto_crtl0, KVMLoongArchPCHPIC),
-        VMSTATE_UINT64(auto_crtl1, KVMLoongArchPCHPIC),
-        VMSTATE_UINT8_ARRAY(route_entry, KVMLoongArchPCHPIC, 64),
-        VMSTATE_UINT8_ARRAY(htmsi_vector, KVMLoongArchPCHPIC, 64),
-        VMSTATE_UINT64(last_intirr, KVMLoongArchPCHPIC),
-        VMSTATE_UINT64(intirr, KVMLoongArchPCHPIC),
-        VMSTATE_UINT64(intisr, KVMLoongArchPCHPIC),
-        VMSTATE_UINT64(int_polarity, KVMLoongArchPCHPIC),
-        VMSTATE_END_OF_LIST()
-    }
-};
-
-
-static void kvm_loongarch_pch_pic_class_init(ObjectClass *oc, void *data)
-{
-    DeviceClass *dc = DEVICE_CLASS(oc);
-    KVMLoongArchPCHPICClass *pch_pic_class = KVM_LOONGARCH_PCH_PIC_CLASS(oc);
-
-    pch_pic_class->parent_realize = dc->realize;
-    dc->realize = kvm_loongarch_pch_pic_realize;
-    pch_pic_class->is_created = false;
-    dc->vmsd = &vmstate_kvm_loongarch_pch_pic;
-
-}
-
-static const TypeInfo kvm_loongarch_pch_pic_info = {
-    .name = TYPE_KVM_LOONGARCH_PCH_PIC,
-    .parent = TYPE_SYS_BUS_DEVICE,
-    .instance_size = sizeof(KVMLoongArchPCHPIC),
-    .class_size = sizeof(KVMLoongArchPCHPICClass),
-    .class_init = kvm_loongarch_pch_pic_class_init,
-};
-
-static void kvm_loongarch_pch_pic_register_types(void)
-{
-    type_register_static(&kvm_loongarch_pch_pic_info);
-}
-
-type_init(kvm_loongarch_pch_pic_register_types)
diff --git a/hw/intc/loongarch_pic_kvm.c b/hw/intc/loongarch_pic_kvm.c
new file mode 100644
index 000000000000..2f977e916e6a
--- /dev/null
+++ b/hw/intc/loongarch_pic_kvm.c
@@ -0,0 +1,38 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/*
+ * LoongArch kvm pch pic interrupt support
+ *
+ * Copyright (C) 2025 Loongson Technology Corporation Limited
+ */
+
+#include "qemu/osdep.h"
+#include "qapi/error.h"
+#include "hw/boards.h"
+#include "hw/intc/loongarch_pch_pic.h"
+#include "hw/loongarch/virt.h"
+#include "hw/pci-host/ls7a.h"
+#include "sysemu/kvm.h"
+
+void kvm_pic_realize(DeviceState *dev, Error **errp)
+{
+    LoongarchPICState *lps = LOONGARCH_PIC(dev);
+    uint64_t pch_pic_base = VIRT_PCH_REG_BASE;
+    int ret;
+
+    ret = kvm_create_device(kvm_state, KVM_DEV_TYPE_LOONGARCH_PCHPIC, false);
+    if (ret < 0) {
+        fprintf(stderr, "Create KVM_LOONGARCH_PCHPIC failed: %s\n",
+                strerror(-ret));
+        abort();
+    }
+
+    lps->dev_fd = ret;
+    ret = kvm_device_access(lps->dev_fd, KVM_DEV_LOONGARCH_PCH_PIC_GRP_CTRL,
+                            KVM_DEV_LOONGARCH_PCH_PIC_CTRL_INIT,
+                            &pch_pic_base, true, NULL);
+    if (ret < 0) {
+        fprintf(stderr, "KVM_LOONGARCH_PCH_PIC_INIT failed: %s\n",
+                strerror(-ret));
+        abort();
+    }
+}
diff --git a/hw/intc/meson.build b/hw/intc/meson.build
index eefe1e357ec7..2596e07ef00d 100644
--- a/hw/intc/meson.build
+++ b/hw/intc/meson.build
@@ -75,8 +75,9 @@ specific_ss.add(when: 'CONFIG_LOONGARCH_IPI', if_true: files('loongarch_ipi.c'))
 specific_ss.add(when: ['CONFIG_KVM', 'CONFIG_LOONGARCH_IPI'],
                 if_true: files('loongarch_ipi_kvm.c'))
 specific_ss.add(when: 'CONFIG_LOONGARCH_PCH_PIC', if_true: files('loongarch_pch_pic.c', 'loongarch_pic_common.c'))
+specific_ss.add(when: ['CONFIG_KVM', 'CONFIG_LOONGARCH_PCH_PIC'],
+                if_true: files('loongarch_pic_kvm.c'))
 specific_ss.add(when: 'CONFIG_LOONGARCH_PCH_MSI', if_true: files('loongarch_pch_msi.c'))
 specific_ss.add(when: 'CONFIG_LOONGARCH_EXTIOI', if_true: files('loongarch_extioi.c', 'loongarch_extioi_common.c'))
 specific_ss.add(when: ['CONFIG_KVM', 'CONFIG_LOONGARCH_EXTIOI'],
                if_true: files('loongarch_extioi_kvm.c'))
-specific_ss.add(when: 'CONFIG_LOONGARCH_PCH_PIC_KVM', if_true: files('loongarch_pch_pic_kvm.c'))
diff --git a/hw/loongarch/Kconfig b/hw/loongarch/Kconfig
index 1c4aa15e6b41..b42a8573d41c 100644
--- a/hw/loongarch/Kconfig
+++ b/hw/loongarch/Kconfig
@@ -14,7 +14,6 @@ config LOONGARCH_VIRT
     select LOONGARCH_PCH_PIC
     select LOONGARCH_PCH_MSI
     select LOONGARCH_EXTIOI
-    select LOONGARCH_PCH_PIC_KVM if KVM
     select LS7A_RTC
     select SMBIOS
     select ACPI_CPU_HOTPLUG
diff --git a/hw/loongarch/virt.c b/hw/loongarch/virt.c
index b0b8c3709515..5bedc6836458 100644
--- a/hw/loongarch/virt.c
+++ b/hw/loongarch/virt.c
@@ -450,26 +450,13 @@ static void virt_irq_init(LoongArchVirtMachineState *lvms)
     lvms->extioi = extioi;
 
     virt_cpu_irq_init(lvms);
-    if (kvm_enabled() && kvm_irqchip_in_kernel()) {
-        pch_pic = qdev_new(TYPE_KVM_LOONGARCH_PCH_PIC);
-        sysbus_realize_and_unref(SYS_BUS_DEVICE(pch_pic), &error_fatal);
-    } else {
-        pch_pic = qdev_new(TYPE_LOONGARCH_PIC);
-        num = VIRT_PCH_PIC_IRQ_NUM;
-        qdev_prop_set_uint32(pch_pic, "pch_pic_irq_num", num);
-        d = SYS_BUS_DEVICE(pch_pic);
-        sysbus_realize_and_unref(d, &error_fatal);
-        memory_region_add_subregion(get_system_memory(), VIRT_IOAPIC_REG_BASE,
-                                sysbus_mmio_get_region(d, 0));
-
-        /* Connect pch_pic irqs to extioi */
-        for (i = 0; i < num; i++) {
-            qdev_connect_gpio_out(DEVICE(d), i, qdev_get_gpio_in(extioi, i));
-        }
-    }
+    pch_pic = qdev_new(TYPE_LOONGARCH_PIC);
+    num = VIRT_PCH_PIC_IRQ_NUM;
+    qdev_prop_set_uint32(pch_pic, "pch_pic_irq_num", num);
+    d = SYS_BUS_DEVICE(pch_pic);
+    sysbus_realize_and_unref(d, &error_fatal);
 
     pch_msi = qdev_new(TYPE_LOONGARCH_PCH_MSI);
-    num = VIRT_PCH_PIC_IRQ_NUM;
     start   =  num;
     num = EXTIOI_IRQS - start;
     qdev_prop_set_uint32(pch_msi, "msi_irq_base", start);
diff --git a/include/hw/intc/loongarch_pch_pic.h b/include/hw/intc/loongarch_pch_pic.h
index 08126efede95..4b52f1165f2e 100644
--- a/include/hw/intc/loongarch_pch_pic.h
+++ b/include/hw/intc/loongarch_pch_pic.h
@@ -10,14 +10,13 @@
 
 #include "hw/intc/loongarch_pic_common.h"
 
-#define TYPE_KVM_LOONGARCH_PCH_PIC      "loongarch_kvm_pch_pic"
-
 #define TYPE_LOONGARCH_PIC  "loongarch_pic"
 #define PCH_PIC_NAME(name)  TYPE_LOONGARCH_PIC#name
 OBJECT_DECLARE_TYPE(LoongarchPICState, LoongarchPICClass, LOONGARCH_PIC)
 
 struct LoongarchPICState {
     LoongArchPICCommonState parent_obj;
+    int dev_fd;
 };
 
 struct LoongarchPICClass {
@@ -27,38 +26,6 @@ struct LoongarchPICClass {
     ResettablePhases parent_phases;
 };
 
-struct KVMLoongArchPCHPIC {
-    SysBusDevice parent_obj;
-    uint64_t int_mask; /*0x020 interrupt mask register*/
-    uint64_t htmsi_en; /*0x040 1=msi*/
-    uint64_t intedge; /*0x060 edge=1 level  =0*/
-    uint64_t intclr; /*0x080 for clean edge int,set 1 clean,set 0 is noused*/
-    uint64_t auto_crtl0; /*0x0c0*/
-    uint64_t auto_crtl1; /*0x0e0*/
-    uint64_t last_intirr;    /* edge detection */
-    uint64_t intirr; /* 0x380 interrupt request register */
-    uint64_t intisr; /* 0x3a0 interrupt service register */
-    /*
-     * 0x3e0 interrupt level polarity selection
-     * register 0 for high level trigger
-     */
-    uint64_t int_polarity;
-
-    uint8_t route_entry[64]; /*0x100 - 0x138*/
-    uint8_t htmsi_vector[64]; /*0x200 - 0x238*/
-};
-typedef struct KVMLoongArchPCHPIC KVMLoongArchPCHPIC;
-DECLARE_INSTANCE_CHECKER(KVMLoongArchPCHPIC, KVM_LOONGARCH_PCH_PIC,
-                         TYPE_KVM_LOONGARCH_PCH_PIC)
-
-struct KVMLoongArchPCHPICClass {
-    SysBusDeviceClass parent_class;
-    DeviceRealize parent_realize;
+void kvm_pic_realize(DeviceState *dev, Error **errp);
 
-    bool is_created;
-    int dev_fd;
-};
-typedef struct KVMLoongArchPCHPICClass KVMLoongArchPCHPICClass;
-DECLARE_CLASS_CHECKERS(KVMLoongArchPCHPICClass, KVM_LOONGARCH_PCH_PIC,
-                       TYPE_KVM_LOONGARCH_PCH_PIC)
 #endif /* HW_LOONGARCH_PCH_PIC_H */
-- 
2.51.0

